bokeh/util/api_crawler.py

```diff
@@ -0,0 +1,232 @@
+
+                    if function_diff:
+                        diff[x]["functions"] = list(function_diff)
+                    else:
+                        diff[x]["functions"] = []
+        return diff
+
+    def diff_methods(self, diff, intersection):
+        for x in intersection.keys():
+            former_classes = self.former[x].get("classes") if self.former.get(x) else {}
+            latter_classes = self.latter[x].get("classes") if self.latter.get(x) else {}
+            if former_classes and latter_classes:
+                for y in intersection[x]["classes"]:
+                    # Prevent NoneType errors by returning empty dict.
+                    former_methods = self.former[x]["classes"].get(y, {})
+                    latter_methods = self.latter[x]["classes"].get(y, {})
+                    if former_methods.get("methods") and latter_methods.get("methods"):
+                        former_values = set(list(former_methods.values())[0])
+                        latter_values = set(list(latter_methods.values())[0])
+                        methods_diff = self._operation(former_values, latter_values)
+                        if methods_diff:
+                            if not diff.get(x):
+                                diff[x] = {}
+                                diff[x]["classes"] = {}
+                            diff[x]["classes"][y] = copy.deepcopy(intersection[x]["classes"][y])
+                            diff[x]["classes"][y]["methods"] = methods_diff
+        return diff
+
+    def diff_modules(self):
+        intersection, diff = self.diff_files()
+        diff = self.diff_functions_classes(diff, intersection)
+        diff = self.diff_methods(diff, intersection)
+        return diff
+
+    def parse_diff(self, diff):
+        parsed_diff = []
+        if self.additions:
+            method = "ADDED"
+        else:
+            method = "DELETED"
+        for x in diff.keys():
+            formatted_string = "%s %s" % (method, os.path.splitext(x.replace("/", "."))[0])
+            if diff[x].values():
+                for y in diff[x].values():
+                    if isinstance(y, dict) and y:
+                        for z in y.keys():
+                            if not y[z].values():
+                                parsed_diff.append("%s.%s" % (formatted_string, z))
+                            else:
+                                for a in y[z].values():
+                                    for b in a:
+                                        parsed_diff.append("%s.%s.%s" % (formatted_string, z, b))
+                    elif isinstance(y, list) and y:
+                        for z in y:
+                            class_string = "%s" % z
+                            parsed_diff.append("%s.%s" % (formatted_string, class_string))
+            else:
+                parsed_diff.insert(0, formatted_string)
+        return parsed_diff
+
+    def get_diff(self):
+        self.additions = False
```

Comment:
The way `additions` and in result `_operation`are handled is cumbersome in my opinion, though as a minor thing I won't ask to rewrite this (it would require a major rewrite of the entire module).